ChiModules/KEigenvalueSolver/IterativeMethods/power_iteration.cc

```diff
@@ -18,168 +18,127 @@ typedef sweep_namespace::SchedulingAlgorithm SchedulingAlgorithm;
-                                    *sweep_chunk);
-
-  // ----- Tool the sweep chunk
-  sweep_chunk->SetDestinationPhi(phi_new_local);
+      << "\n\n********** Solving k-eigenvalue problem with "
+      << "the Power Method.\n\n";
 
-  // ----- Set starting guess to a unit magnitude flux
-  phi_prev_local.assign(phi_prev_local.size(),1.0);
-  ScopedCopySTLvectors(groupset, phi_prev_local, phi_old_local);
+  phi_old_local.assign(phi_old_local.size(), 1.0);
 
-  // ----- Start outer k iterations
-  double F_prev = 1.0;       //Production source prev
-  double k_eff_prev = k_eff;
-  int nit = 0;               //number of iterations
-  bool k_converged = false;
+  double F_prev = 1.0;
+  double k_eff_prev = 1.0;
+  double k_eff_change = 1.0;
 
+  //================================================== Start power iterations
+  int nit = 0;
+  bool converged = false;
   while (nit < options.max_iterations)
   {
-    chi_log.Log(LOG_0VERBOSE_2)
-      << "\n********** Starting source iterations";
-
-
-    // ----- Start inner source iterations
-    double pw_change_prev = 1.0;
-    bool si_converged = false;
-    for (int si_nit=0; si_nit<groupset.max_iterations; si_nit++)
+    //============================================= Loop over groupsets
+    MPI_Barrier(MPI_COMM_WORLD);
+    int gs = -1;
+    for (auto& groupset : group_sets)
     {
-      // ----- Set source and sweep
-      SetKSource(groupset, APPLY_MATERIAL_SOURCE |
-                           APPLY_AGS_SCATTER_SOURCE | APPLY_WGS_SCATTER_SOURCE |
-                           APPLY_FISSION_SOURCE);
-      groupset.angle_agg.ZeroOutgoingDelayedPsi();
-      phi_new_local.assign(phi_new_local.size(),0.0);
-      SweepScheduler.Sweep();
-
-      // ----- Compute convergence parameters
-      double pw_change = ComputePiecewiseChange(groupset);
-      ScopedCopySTLvectors(groupset, phi_new_local, phi_old_local);
-      double rho = sqrt(pw_change/pw_change_prev);
-      pw_change_prev = pw_change;
-      nit += 1;
-
-      if (si_nit==0) rho = 0.0;
-      if (pw_change<std::max(groupset.residual_tolerance*rho,1.0e-10))
-        si_converged = true;
-
-      // ----- Print iteration information
-      std::string offset = "    ";
-      std::stringstream si_iter_info;
-      si_iter_info
-        << chi_program_timer.GetTimeString() << " "
-        << offset
-        << "WGS groups ["
-        << groupset.groups.front().id
-        << "-"
-        << groupset.groups.back().id
-        << "]"
-        << " Source Iteration " << std::setw(5) << si_nit
-        << " Point-wise change " << std::setw(14) << pw_change;
-        
-      if (si_converged)
-        si_iter_info << " CONVERGED\n";
-      chi_log.Log(LOG_0VERBOSE_1) << si_iter_info.str();
-
-      if (si_converged) {
-        chi_log.Log(LOG_0VERBOSE_1)
-            << "\nSource iterations converged in "
-            << si_nit << " iteratrions.\n";
-        break;
+      ComputeSweepOrderings(groupset);
```

Comment:
Can we not store sweep orderings?  If not, we should probably create an issue.  Storage of the orderings should be very lightweight compared to the computation.

Reply:
Tried this before and it can get quite beefy when lots of angles are involved. Here's what happens:
- Let's say 10,000 cells on a process
- The order of the cells in the topo-sort alone will require 10,000 x 8-byte ids
- Inter-location dependencies, cycles, location-to-location sweep ordering, all take additional overhead space but lets say we have to save an additional 1.5 times the amount of cells in "overhead"
- Let's say 5000 angles in a groupset

Thats 10,000 x 1.5 x 5,000 x 8 bytes = 600 Mb.

Times 3 groupsets, prestored? 1.8 Gb. Ouch.

Reply:
Interesting.  Having just been through all of this with the GPU sweeper, I think we can save a lot of memory here.  At least half of non-overhead, and a pretty good percentage of overhead by smartly picking anglesets to reduce the number of cycles. I'll create an issue to remind us to look at this.

Reply:
It just dawned on me that you say 5000 angles in a groupset.  Unless I misunderstand, that number should be much, much lower unless we're using single angle aggregation (which wouldn't make much sense since we included considerable overhead for cycles).  Instead of 5000, shouldn't that number be on the order of 8 to a few hundred?

Reply:
Poor @zachhardy  got caught up in a design/architecture war here.  :)  To keep from holding him up,  I'm going to create separate issues for this and the MPI barrier work.